lsm.go

```diff
@@ -0,0 +1,352 @@
+			}
+		}
+	}
+
+	return nil
+}
+
+func (l *LSM) MaxLevel() SentinelType {
+	return SentinelType(len(l.configs) - 1)
+}
+
+func (l *LSM) Add(tx uint64, record arrow.Record) {
+	record.Retain()
+	size := util.TotalRecordSize(record)
+	l.levels.Prepend(parts.NewArrowPart(tx, record, int(size), nil))
+	l0 := l.sizes[L0].Add(int64(size))
+	l.metrics.LevelSize.WithLabelValues(L0.String()).Set(float64(l0))
+	if l0 >= l.configs[L0].MaxSize {
+		go l.compact()
+	}
+}
+
+// InsertPart inserts a part into the LSM tree. It will be inserted into the correct level. It does not check if the insert should cause a compaction.
+// This should only be used during snapshot recovery.
+func (l *LSM) InsertPart(level SentinelType, part *parts.Part) {
+	l.findLevel(level).Prepend(part)
+}
+
+func (l *LSM) String() string {
+	return l.levels.String()
+}
+
+func (l *LSM) Prefixes(ctx context.Context, prefix string) ([]string, error) {
+	return []string{l.prefix}, nil
+}
+
+func (l *LSM) Scan(ctx context.Context, _ string, _ *dynparquet.Schema, filter logicalplan.Expr, tx uint64, callback func(context.Context, any) error) error {
+	l.RLock()
+	defer l.RUnlock()
+
+	booleanFilter, err := BooleanExpr(filter)
+	if err != nil {
+		return fmt.Errorf("boolean expr: %w", err)
+	}
+	var iterError error
+	l.levels.Iterate(func(node *Node) bool {
+		if node.part == nil { // encountered a sentinel node; continue on
+			return true
+		}
+
+		if node.part.TX() > tx { // skip parts that are newer than this transaction
+			return true
+		}
+
+		if r := node.part.Record(); r != nil {
+			r.Retain()
+			if err := callback(ctx, r); err != nil {
+				iterError = err
+				return false
+			}
+			return true
+		}
+
+		buf, err := node.part.AsSerializedBuffer(nil)
+		if err != nil {
+			iterError = err
+			return false
+		}
+
+		for i := 0; i < buf.NumRowGroups(); i++ {
+			rg := buf.DynamicRowGroup(i)
+			mayContainUsefulData, err := booleanFilter.Eval(rg)
+			if err != nil {
+				iterError = err
+				return false
+			}
+
+			if mayContainUsefulData {
+				if err := callback(ctx, rg); err != nil {
+					iterError = err
+					return false
+				}
+			}
+		}
+		return true
+	})
+	return iterError
+}
+
+// TODO: this should be changed to just retain the sentinel nodes in the lsm struct to do an O(1) lookup
+func (l *LSM) findLevel(level SentinelType) *Node {
+	var list *Node
+	l.levels.Iterate(func(node *Node) bool {
+		if node.part == nil && node.sentinel == level {
+			list = node
+			return false
+		}
+		return true
+	})
+
+	return list
+}
+
+// Merge will merge the given level into an arrow record for the next level using the configured Compact function for the given level.
+// If this is the max level of the LSM an external writer must be provided to write the merged part elsewhere.
+func (l *LSM) merge(level SentinelType, externalWriter func([]*parts.Part) (*parts.Part, int64, int64, error)) error {
```

Comment:
What use cases does the externalWriter cover?

Reply:
For the record I hate this interface and want to refactor this out. But right now the externalWriter is for serialization during persistence.